eng/pipelines/templates/steps/sync-repo-branch.yml

```diff
@@ -1,17 +1,97 @@
 parameters:
-  Repos: # Provde a list of repos
+  # Provde a list of repos
+  # Example:
+  #   Azure/azure-rest-api-specs:
+  #     Branch: master
+  #     TargetRepos: 
+  #       Azure/azure-rest-api-specs-pr:
+  #       azure-sdk/azure-rest-api-specs:
+  #         Branch: dev
+  #         Rebase: true
+  #   Azure/azure-sdk-for-go:
+  #     Branch: master
+  #     TargetRepos: 
+  #       Azure/azure-sdk-for-go-pr:
+  - name: Repos
+    type: object
+  # Github PAT
+  - name: GH_TOKEN
+    type: string
 
 steps:
 
 - ${{ each repo in parameters.Repos }}:
-  - script: |
-      git clone --filter=blob:none --branch ${{ repo.value.Branch }} https://github.com/${{ repo.key }} ${{ repo.key }}
+  - pwsh: |
+      Set-PsDebug -Trace 1
+      $SourceRepo = '${{ repo.key }}'
+      $SourceBranch = '${{ repo.value.Branch }}'
+
+      if (-not (Test-Path ${{ repo.key }})) {
+        New-Item -Path ${{ repo.key }} -ItemType Directory -Force
+        Set-Location $SourceRepo
+        git init
+        git remote add Source "https://${{ parameters.GH_TOKEN }}@github.com/${SourceRepo}.git"
+      } else {
+        Set-Location $SourceRepo
+      }
+
+      git fetch --no-tags Source $SourceBranch
+      if ($LASTEXITCODE -ne 0) {
+        Write-Host "##vso[task.logissue type=error]Failed to fetch ${SourceRepo}:${SourceBranch}"
+        exit 1
+      }
+
+      git checkout -B source_branch "refs/remotes/Source/${SourceBranch}"
+
+      Set-PsDebug -Off
+
     displayName: ${{ repo.key }} - Clone ${{ repo.value.Branch }}
     continueOnError: true
 
   - ${{ each target in repo.value.TargetRepos }}:
-    - script: | 
-        git push https://$(azuresdk-github-pat)@github.com/${{ target.key }} ${{ coalesce(target.value.Branch, repo.value.Branch) }}
+    - pwsh: |
```

Comment:
Given that this script is pretty sizable can you please pull it out and put it into a script under `eng/scripts` and all it from this step? It will allow us to call this locally to try and repro any issues we might hit and we might be able to reuse it in other contexts.

Reply:
I'm wondering if this will still work out of this repo with template reference?